libmamba/src/solv-cpp/pool.hpp

```diff
@@ -0,0 +1,214 @@
+ *******************************/
+
+#include <solv/pool.h>
+
+namespace mamba::solv
+{
+
+    template <typename UnaryFunc>
+    void ObjPool::for_each_repo_id(UnaryFunc func) const
+    {
+        const ::Pool* const pool = raw();
+        const ::Repo* repo = nullptr;
+        RepoId repo_id = 0;
+        FOR_REPOS(repo_id, repo)
+        {
+            func(repo_id);
+        }
+    }
+
+    template <typename UnaryFunc>
+    void ObjPool::for_each_repo(UnaryFunc func) const
+    {
+        // Safe optional unchecked because we iterate over available values
+        return for_each_repo_id([this, func](RepoId id) { func(get_repo(id).value()); });
+    }
+
+    template <typename UnaryFunc>
+    void ObjPool::for_each_repo(UnaryFunc func)
+    {
+        // Safe optional unchecked because we iterate over available values
+        return for_each_repo_id([this, func](RepoId id) { func(get_repo(id).value()); });
+    }
+
+    template <typename UnaryFunc>
+    void ObjPool::for_each_whatprovides_id(DependencyId dep, UnaryFunc func) const
+    {
+        auto* const pool = const_cast<::Pool*>(raw());
+        SolvableId id = 0;
+        ::Id offset = 0;  // Not really an Id
+        FOR_PROVIDES(id, offset, dep)
+        {
+            func(id);
+        }
+    }
+
+    template <typename UnaryFunc>
+    void ObjPool::for_each_whatprovides(DependencyId dep, UnaryFunc func) const
+    {
+        // Safe optional unchecked because we iterate over available values
+        return for_each_whatprovides_id(
+            dep,
+            [this, func](SolvableId id) { func(get_solvable(id)).value(); }
+        );
+    }
+
+    template <typename UnaryFunc>
+    void ObjPool::for_each_whatprovides(DependencyId dep, UnaryFunc func)
+    {
+        // Safe optional unchecked because we iterate over available values
+        return for_each_whatprovides_id(
+            dep,
+            [this, func](SolvableId id) { func(get_solvable(id).value()); }
+        );
+    }
+
+    template <typename UnaryFunc>
+    void ObjPool::for_each_solvable_id(UnaryFunc func) const
+    {
+        const ::Pool* const pool = raw();
+        SolvableId id = 0;
+        FOR_POOL_SOLVABLES(id)
+        {
+            func(id);
+        }
+    }
+
+    template <typename UnaryFunc>
+    void ObjPool::for_each_solvable(UnaryFunc func) const
+    {
+        // Safe optional unchecked because we iterate over available values
+        return for_each_solvable_id([this, func](SolvableId id) { func(get_solvable(id).value()); });
+    }
+
+    template <typename UnaryFunc>
+    void ObjPool::for_each_solvable(UnaryFunc func)
+    {
+        // Safe optional unchecked because we iterate over available values
+        return for_each_solvable_id([this, func](SolvableId id) { func(get_solvable(id).value()); });
+    }
+
+    template <typename Func>
+    void ObjPool::set_debug_callback(Func callback)
+    {
+        m_user_debug_callback.reset(new Func(std::move(callback)));
```

Comment:
Did you consider `m_user_debug_callback` as `std::function<...>` and passing a pointer to it in `pool_set_debug_callback` instead?
It looks like `m_user_debug_callback` is passed as data to `pool_set_debug_callback` and we have the `debug_callback` reinterpreting the data anyway, so we dont have to store the real callback in a complicated way. Am I missing something?

Reply:
I did, but passing the pointer to an attribute does not play well with move semantics.
I could try something with `std::function::target`, but I am also not sure it is heap allocated.

Reply:
Ok moving would imply re-setting the callback on move, so it's either that or having this allocation.
Let's keep it like that.